src/java/org/apache/cassandra/index/sai/view/View.java

```diff
@@ -0,0 +1,124 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.view;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.SSTableIndex;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.utils.Interval;
+import org.apache.cassandra.utils.IntervalTree;
+
+public class View implements Iterable<SSTableIndex>
+{
+    private final Map<Descriptor, SSTableIndex> view;
+
+    private final TermTree termTree;
+    private final AbstractType<?> keyValidator;
+    private final IntervalTree<Key, SSTableIndex, Interval<Key, SSTableIndex>> keyIntervalTree;
+
+    public View(IndexContext context, Collection<SSTableIndex> indexes)
+    {
+        this.view = new HashMap<>();
+        this.keyValidator = context.keyValidator();
+
+        AbstractType<?> validator = context.getValidator();
+
+        TermTree.Builder termTreeBuilder = new RangeTermTree.Builder(validator);
+
+        List<Interval<Key, SSTableIndex>> keyIntervals = new ArrayList<>();
+        for (SSTableIndex sstableIndex : indexes)
+        {
+            this.view.put(sstableIndex.getSSTable().descriptor, sstableIndex);
+            termTreeBuilder.add(sstableIndex);
+            keyIntervals.add(Interval.create(new Key(sstableIndex.minKey()),
+                                             new Key(sstableIndex.maxKey()),
+                                             sstableIndex));
+        }
+
+        this.termTree = termTreeBuilder.build();
+        this.keyIntervalTree = IntervalTree.build(keyIntervals);
+    }
+
+    public Set<SSTableIndex> match(Expression expression)
+    {
+        return termTree.search(expression);
+    }
+
+    public List<SSTableIndex> match(DecoratedKey minKey, DecoratedKey maxKey)
+    {
+        return keyIntervalTree.search(Interval.create(new Key(minKey), new Key(maxKey), null));
+    }
+
+    public Iterator<SSTableIndex> iterator()
+    {
+        return view.values().iterator();
+    }
+
+    public Collection<SSTableIndex> getIndexes()
+    {
+        return view.values();
+    }
+
+    public boolean containsSSTable(SSTableReader sstable)
+    {
+        return view.containsKey(sstable.descriptor);
+    }
+
+    public int size()
+    {
+        return view.size();
+    }
+
+    /**
+     * This is required since IntervalTree doesn't support custom Comparator
+     * implementations and relied on items to be comparable which "raw" keys are not.
```

Comment:
`DecoratedKey` is comparable though, right? (Tried it, and some of the basic integration tests look like they're still passing...)

Reply:
I can only assume that at some point in the past this was a problem. I've removed this and will see if it throws any problems during testing.